### ISSUE TITLE
Treat EDN nil as Go zero value

### ISSUE THREAD

**Issue by tolitius** [go-edn/edn#11]
some systems that I don't control could send:

```clojure
{:a nil :b 34}
```

which would result in:

```
edn: cannot unmarshal nil into Go value of type string.
```

it would make sense to convert EDN `nil`s to a Go type [zero values](https://tour.golang.org/basics/12), or simply drop them. Then we can either not worry about it or use `",omitempty"` to ignore them.

What do you think?

**Comment by tolitius** [comment]
i.e. why not:
```go
v.Set(reflect.Zero(v.Type()))
```
here: https://github.com/go-edn/edn/blob/470b0ab438bbcd34ad4b5cab3d1eb1c13d69d6b9/decode.go#L910-L912
?

**Comment by hypirion** [comment]
It depends on how you need to treat `nil`s. Are they equal to the zero value in your case, or are they identical to an absence of value?

As a short term solution, you can use
``` go
type Good struct {
  A *string
  B int
}
```
instead of
```go
type Error struct {
  A string
  B int
}
```

And also get information about whether this is a `nil` value or not.

---

So in general, I think there's value in distinguishing the between the zero value (`""`, `0`, etc) `nil`, and the absence of a value. Consider the following patch bodies:

```clj
;; 1
{:author #person/id 0}
;; 2
{:author nil}
;; 3
{}
```

Typical behaviour is that the first one sets the author to the zero value, the second sets the author to null, and the third does not update the author field.

I'm not against `nil` being equal to the zero value or the absence of a value, but I don't think I'd like it to be the default. Maybe a new option `ignore-nil` or `nil-as-zero` or something? I need to think about what makes the most sense here, as ignoring the value is different from setting it to the zero value.

**Comment by tolitius** [comment]
> _there's value in distinguishing the between the zero value ("", 0, etc) nil, and the absence of a value_

I agree, although (talking about unmarshalling EDN) I feel that EDN's nil (on a Clojure side) most of the time means absence of the value. i.e. `{:author #person/id 0}` does not mean nil in Clojure.

> _Maybe a new option `ignore-nil` or `nil-as-zero` or something?_

I think both may have a place.

If you decide to leave the current behavior as a default one (which you are leaning to) it would help to have a mechanism to define `ignore-nil` or `nil-as-zero` on a struct level, otherwise it would go back to `required vs. optional` protobuf (2.x) problem. (in protobuf 3 fields are optional by default).

**Comment by tolitius** [comment]
[the commit above]: me trying things, i.e. no intention to change the official repo

Rhetorical observation: 

```go
	wantptr := ttype == tokenSymbol && bytes.Equal(nilByte, bs)
```

When unmarshalling from `EDN` I don't think `nilByte` (i.e. EDN's `nil`) means "want a Go pointer", but rather "nothing" for whatever type.

**Comment by hypirion** [comment]
Yeah, the idea is to have `ignore-nil` or what we decide to call it a property in struct tags, e.g.

```go
type Foo struct {
    Bar string `edn:",ignore-nil"
}
```
